### PR TITLE
Function Contracts: Interior Mutability Tests

### PR DESCRIPTION
**Pull request by pi314mm**
Test cases for interior mutability using Cell, OnceCell, UnsafeCell, and RefCell.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by feliperodri — CHANGES_REQUESTED** [review]
We also need to update our RFC to describe any limitations around interior mutability.

File: tests/expected/function-contract/interior-mutability/cell.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#[kani::requires(*unsafe{x.x.hack()} < 100)]
+#[kani::modifies(x.x.hack())]
+#[kani::ensures(|_| *unsafe{x.x.hack()} < 101)]
+fn modify(x: &InteriorMutability) {
+    x.x.set(x.x.get() + 1)
+}
+
+#[kani::proof_for_contract(modify)]
+fn main() {
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/cell.rs:44]: Please provide the harness a more meaningful name.

File: rfc/src/rfcs/0009-function-contracts.md [hunk 2]
```diff
@@ -706,6 +706,22 @@ structures however are ubiquitous and users can avoid the unsoundness with
 relative confidence by overprovisioning (generating inputs that are several
 times larger than what they expect the function will touch).
 
+### Interior Mutability
```
Comment (celinval) [inline rfc/src/rfcs/0009-function-contracts.md:709]: I don't know if this should be here. This is not related to contracts. It may be more interesting to put something like this in our wiki. Maybe in the [limitations](https://model-checking.github.io/kani/limitations.html) section.

File: tests/expected/function-contract/interior-mutability/cell.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    unsafe fn expose(&self) -> &T {
+        transmute(self)
+    }
+}
+
+// ---------------------------------------------------
+//                      Test Case
+// ---------------------------------------------------
+
+// This struct is contains Cell which can be mutated
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/cell.rs:30]: use `///` for doc comments

File: tests/expected/function-contract/interior-mutability/cell_stub.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+use std::cell::Cell;
+use std::mem::transmute;
+
+/// This exposes the underlying representation so that it can be added into a modifies clause within kani
+trait Exposeable<T: ?Sized> {
+    unsafe fn expose(&self) -> &T;
+}
+
+// This unsafe manipulation is valid due to Cell having the same underlying data layout as its internal T as explained here: https://doc.rust-lang.org/stable/std/cell/struct.Cell.html#memory-layout
+impl<T: ?Sized> Exposeable<T> for Cell<T> {
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/cell_stub.rs:21]: Can you use `Cell::get` or `Cell::as_ptr` instead?

**Review by celinval — COMMENTED** [review]
I don't think you need the `transmute` in any of these cases. You should be able to use the public methods to get the data address or a copy of the data.

We have two problems, and I think it might be helpful to think about them separately, although you might want to add a test that has both. The first problem is interior mutability, and the second one is encapsulation.

Using `transmute` is necessary if you want to break encapsulation. The interior mutability could be a problem for identifying expressions that are side-effect free.

File: tests/expected/function-contract/interior-mutability/cell_stub.rs (hunk 4 above)
Comment (pi314mm) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: This can be done for `Cell` and `UnsafeCell` and I've added two more test cases for this. I don't immediately see a way to do this for `OnceCell`. It won't extend to `RefCell` due to runtime borrow checks.

Comment (celinval) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: You can use the address of the `OnceCell` itself.

File: tests/expected/function-contract/interior-mutability/transmute/oncecell.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/// This struct is contains OnceCell which can be mutated
+struct InteriorMutability {
+    x: OnceCell<u32>,
+}
+
+/// contracts need to access im.x internal data through the unsafe function im.x.expose()
+#[kani::requires(unsafe{im.x.expose()}.is_none())]
+#[kani::modifies(im.x.expose())]
+#[kani::ensures(|_| unsafe{im.x.expose()}.is_some())]
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/transmute/oncecell.rs:37]:
```suggestion
#[kani::requires(im.x.get().is_none())]
#[kani::modifies(&im.x)]
#[kani::ensures(|_| im.x.get().is_some())]
```

File: tests/expected/function-contract/interior-mutability/cell_stub.rs (hunk 4 above)
Comment (pi314mm) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: Using the address of `OnceCell` itself works! `RefCell::as_ptr()` would let you get the reference to the private field of `self.value` but it won't let you get the other fields `self.borrow` and `self.borrowed_at` that also are modified.

Comment (celinval) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: In that case, you should also use the address of the `RefCell` object.

Comment (pi314mm) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: Using the address of the `OneCell` or the `RefCell` will work for instrumenting the `check` generated function, but will not work for the `replace` generated function because they do not implement the `Arbitrary` trait, nor can they without breaking encapsulation. The current tests that exist can be replaced by a contract stub via `stub_verified`.

Comment (celinval) [reply tests/expected/function-contract/interior-mutability/cell_stub.rs]: Yes. Kani cannot stub contracts that do modifies variables of types that do not implement Arbitrary type. This is an orthogonal problem to interior mutability and to encapsulation.

It's OK if you want to add a test that shows how to handle all three problems at the same time; however, I would ask you to update the test documentation to make this clear.

File: tests/expected/function-contract/interior-mutability/api/unsafecell.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// kani-flags: -Zfunction-contracts
+
+use std::cell::UnsafeCell;
+
+/// This struct is contains UnsafeCell which can be mutated
+struct InteriorMutability {
+    x: UnsafeCell<u32>,
+}
+
+/// contracts need to access im.x internal data through im.x.get() with an unsafe raw pointer dereference
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/api/unsafecell.rs:12]: please update

File: tests/expected/function-contract/interior-mutability/transmute/cell.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of a Cell used as interior mutability in an immutable struct
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/transmute/cell.rs:5]: This test does not need transmute and it is already covered by the other `cell.rs` test.

File: tests/expected/function-contract/interior-mutability/transmute/cell_stub.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to show that the contracts for double can be replaced as a stub within the contracts for quadruple.
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/transmute/cell_stub.rs:5]: cell stub also does not need transmute, so please move it out of this folder.

File: tests/expected/function-contract/interior-mutability/transmute/oncecell.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of an OnceCell used as interior mutability in an immutable struct
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/transmute/oncecell.rs:5]: This test is only using contract for checking, so it also does not require transmute. Please move.

That said, you could add a similar test that instead is written with transmute to enable stubbing. I would highly discourage doing so, since there's no guarantee about the OnceCell layout. If you do write one, please add documentation saying this is just done for test purpose.

File: tests/expected/function-contract/interior-mutability/transmute/unsafecell.rs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of an UnsafeCell used as interior mutability in an immutable struct
```
Comment (celinval) [inline tests/expected/function-contract/interior-mutability/transmute/unsafecell.rs:5]: this does not need transmute

File: tests/expected/function-contract/interior-mutability/api/unsafecell.rs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+use std::cell::UnsafeCell;
+
+/// This struct is contains UnsafeCell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/unsafecell.rs:7]:
```suggestion
/// This struct contains UnsafeCell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/api/unsafecell.rs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+struct InteriorMutability {
+    x: UnsafeCell<u32>,
+}
+
+#[kani::requires(unsafe{*im.x.get()} < 100)]
+#[kani::modifies(im.x.get())]
+#[kani::ensures(|_| unsafe{*im.x.get()} < 101)]
+///im is an immutable reference with interior mutability
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/unsafecell.rs:15]:
```suggestion
/// `im` is an immutable reference with interior mutability
```

File: tests/expected/function-contract/interior-mutability/api/cell_stub.rs [hunk 13]
```diff
@@ -0,0 +1,40 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to show that the contracts for double can be replaced as a stub within the contracts for quadruple.
+/// This shows that we can generate kani::any() for Cell
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/cell_stub.rs:6]:
```suggestion
/// This shows that we can generate `kani::any()` for Cell.
```

File: tests/expected/function-contract/interior-mutability/api/cell_stub.rs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+use std::cell::Cell;
+
+/// This struct is contains Cell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/cell_stub.rs:9]:
```suggestion
/// This struct contains Cell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/api/cell.rs [hunk 15]
```diff
@@ -0,0 +1,25 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// Mutating Cell via as_ptr in contracts
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/cell.rs:5]:
```suggestion
/// Mutating Cell via `as_ptr` in contracts
```

File: tests/expected/function-contract/interior-mutability/api/cell.rs [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+use std::cell::Cell;
+
+/// This struct is contains Cell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/api/cell.rs:8]:
```suggestion
/// This struct contains Cell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/whole-struct/cell.rs [hunk 17]
```diff
@@ -0,0 +1,25 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of a Cell used as interior mutability in an immutable struct
+use std::cell::Cell;
+
+/// This struct is contains Cell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/cell.rs:8]:
```suggestion
/// This struct contains Cell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/whole-struct/cell.rs [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+struct InteriorMutability {
+    x: Cell<u32>,
+}
+
+#[kani::requires(im.x.get() < 100)]
+#[kani::modifies(&im.x)]
+#[kani::ensures(|_| im.x.get() < 101)]
+///im is an immutable reference with interior mutability
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/cell.rs:16]:
```suggestion
/// `im` is an immutable reference with interior mutability
```

File: tests/expected/function-contract/interior-mutability/whole-struct/oncecell.rs [hunk 19]
```diff
@@ -0,0 +1,24 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of an OnceCell used as interior mutability in an immutable struct
+use std::cell::OnceCell;
+
+/// This struct is contains OnceCell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/oncecell.rs:8]:
```suggestion
/// This struct contains OnceCell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/whole-struct/refcell.rs [hunk 20]
```diff
@@ -0,0 +1,25 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of a RefCell used as interior mutability in an immutable struct
+use std::cell::RefCell;
+
+/// This struct is contains Cell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/refcell.rs:8]:
```suggestion
/// This struct contains Cell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/whole-struct/refcell.rs [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+struct InteriorMutability {
+    x: RefCell<u32>,
+}
+
+#[kani::requires(unsafe{*im.x.as_ptr()} < 100)]
+#[kani::modifies(&im.x)]
+#[kani::ensures(|_| unsafe{*im.x.as_ptr()} < 101)]
+///im is an immutable reference with interior mutability
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/refcell.rs:16]:
```suggestion
/// `im` is an immutable reference with interior mutability
```

File: tests/expected/function-contract/interior-mutability/whole-struct/unsafecell.rs [hunk 22]
```diff
@@ -0,0 +1,25 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts
+
+/// The objective of this test is to check the modification of an UnsafeCell used as interior mutability in an immutable struct
+use std::cell::UnsafeCell;
+
+/// This struct is contains UnsafeCell which can be mutated
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/unsafecell.rs:8]:
```suggestion
/// This struct contains UnsafeCell which can be mutated
```

File: tests/expected/function-contract/interior-mutability/whole-struct/unsafecell.rs [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+struct InteriorMutability {
+    x: UnsafeCell<u32>,
+}
+
+#[kani::requires(unsafe{*im.x.get()} < 100)]
+#[kani::modifies(im.x.get())]
+#[kani::ensures(|_| unsafe{*im.x.get()} < 101)]
+///im is an immutable reference with interior mutability
```
Comment (feliperodri) [inline tests/expected/function-contract/interior-mutability/whole-struct/unsafecell.rs:16]:
```suggestion
/// `im` is an immutable reference with interior mutability
```

**Review by feliperodri — APPROVED** [review]
